op-supervisor/supervisor/backend/db/db.go

```diff
@@ -0,0 +1,494 @@
+// event-hash: H(origin, timestamp, payloadhash); enough to check identifier matches & payload matches.
+type DB struct {
+	log    log.Logger
+	m      Metrics
+	data   dataAccess
+	rwLock sync.RWMutex
+
+	lastEntryIdx     int64
+	lastEntryContext logContext
+}
+
+func NewFromFile(logger log.Logger, m Metrics, path string) (*DB, error) {
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0o666)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open database at %v: %w", path, err)
+	}
+	info, err := file.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("failed to stat database at %v: %w", path, err)
+	}
+	lastEntryIdx := info.Size()/entrySize - 1
+	db := &DB{
+		log:          logger,
+		m:            m,
+		data:         file,
+		lastEntryIdx: lastEntryIdx,
+	}
+	if err := db.init(); err != nil {
+		return nil, fmt.Errorf("failed to init database: %w", err)
+	}
+	return db, nil
+}
+
+func (db *DB) init() error {
+	db.updateEntryCountMetric()
+	if db.lastEntryIdx < 0 {
+		// Database is empty so no context to load
+		return nil
+	}
+	lastCheckpoint := (db.lastEntryIdx / searchCheckpointFrequency) * searchCheckpointFrequency
+	i, err := db.newIterator(lastCheckpoint)
+	if err != nil {
+		return fmt.Errorf("failed to create iterator at last search checkpoint: %w", err)
+	}
+	// Read all entries until the end of the file
+	for {
+		_, _, _, err := i.NextLog()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return fmt.Errorf("failed to init from existing entries: %w", err)
+		}
+	}
+	db.lastEntryContext = logContext{
+		blockNum: i.blockNum,
+		logIdx:   i.logIdx,
+	}
+	return nil
+}
+
+func (db *DB) updateEntryCountMetric() {
+	db.m.RecordEntryCount(db.lastEntryIdx + 1)
+}
+
+// ClosestBlockInfo returns the block number and hash of the highest recorded block at or before blockNum.
+// Since block data is only recorded in search checkpoints, this may return an earlier block even if log data is
+// recorded for the requested block.
+func (db *DB) ClosestBlockInfo(blockNum uint64) (uint64, TruncatedHash, error) {
+	db.rwLock.RLock()
+	defer db.rwLock.RUnlock()
+	checkpointIdx, err := db.searchCheckpoint(blockNum, math.MaxUint32)
+	if err != nil {
+		return 0, TruncatedHash{}, fmt.Errorf("no checkpoint at or before block %v found: %w", blockNum, err)
+	}
+	checkpoint, err := db.readSearchCheckpoint(checkpointIdx)
+	if err != nil {
+		return 0, TruncatedHash{}, fmt.Errorf("failed to reach checkpoint: %w", err)
+	}
+	canonicalHash, err := db.readCanonicalHash(checkpointIdx + 1)
+	if err != nil {
+		return 0, TruncatedHash{}, fmt.Errorf("failed to read canonical hash: %w", err)
+	}
+	return checkpoint.blockNum, canonicalHash, nil
+}
+
+// Contains return true iff the specified logHash is recorded in the specified blockNum and logIdx.
+// logIdx is the index of the log in the array of all logs the block.
+func (db *DB) Contains(blockNum uint64, logIdx uint32, logHash common.Hash) (bool, error) {
```

Comment:
I think we should use the dedicated type for the truncated log hash. It's not a hash of the log-payload, but also includes some identifier information (like contract address).

And if we do a lookup of an executing link later, we'll also only have the truncated version, so then it's nice to provide that as argument to this function.

Reply:
Yeah I was originally hoping to encapsulate the truncation (I don't think we ever need to return an initiating message log hash - only the executing message) but it leaks out with executing messages anyway so may as well leave the caller to handle all truncation.

Reply:
Done.  We're still getting the block hash in as a common.Hash mostly because the Add API takes a `eth.BlockID` but we can change that based on what winds up making most sense to the caller.